FWCore/Concurrency/interface/WaitingTask.h

```diff
@@ -48,32 +48,41 @@ namespace edm {
     ///Returns exception thrown by dependent task
     /** If the value evalutes to true then the dependent task failed.
     */
-    std::exception_ptr const& exceptionPtr() const { return m_ptr; }
+    std::exception_ptr exceptionPtr() const {
+      if (m_ptrSet == static_cast<unsigned char>(State::kSet)) {
+        return m_ptr;
+      }
+      return std::exception_ptr{};
+    }
+
+  protected:
+    std::exception_ptr const& uncheckedExceptionPtr() const { return m_ptr; }
 
   private:
+    enum class State : unsigned char { kUnset = 0, kSetting = 1, kSet = 2 };
     ///Called if waited for task failed
     /**Allows transfer of the exception caused by the dependent task to be
      * moved to another thread.
      * This method should only be called by WaitingTaskList
      */
     void dependentTaskFailed(std::exception_ptr iPtr) {
-      bool isSet = false;
-      if (iPtr and m_ptrSet.compare_exchange_strong(isSet, true)) {
+      unsigned char isSet = static_cast<unsigned int>(State::kUnset);
```

Comment:
In general a nice solution to the overall issue. Are the static_cast's necessary? isSet and the enum are both already unsigned char.

Reply:
@wddgit it wouldn't compile without the static_casts.

Reply:
I think the explicit cast is needed because of `enum class` (with plain `enum` cast should not be necessary). But why not cast via `unsigned char` as both source and destination are only that wide?